### ISSUE TITLE
njit in histogram

### ISSUE THREAD

**Issue by wesley34** [numba/numba#8514]
I have wrote my questions here. Let me repost here again
https://stackoverflow.com/questions/74077065/numba-histogram-vs-numpy-histogram

`@njit
def hist1d(x):
   return np.histogram(x)[0]

x = np.array([1/2,1/2,1/2,1/2])
print(hist1d(x),np.histogram(x)[0])`

output: array([0,0,0,0,0,0,0,0,0,0]),array([0,0,0,0,0,0,4,0,0,0])

The output of these functions are different. For the other case should be ok whenever the elements inside matrix are not the same. It seems that there are some bugs in numba on np.histogram.

Thanks~

**Comment by guilhermeleobas** [comment]
Hi @wesley34, thanks for the report. I can confirm this is a bug in Numba.

```python
import numpy as np
from numba import njit

@njit
def hist1d(x):
   return np.histogram(x)[0]


x = np.array([1/2,1/2,1/2,1/2])
print(hist1d(x))
print(hist1d.py_func(x))
```

The bug is in this part of Numba impl. of `np.histogram`, which only fills the `hist` array if `bin_max > bin_min`.
https://github.com/numba/numba/blob/f4674f98795d0c1bbeba693a075885c2e1eace06/numba/np/arraymath.py#L3841-L3851

**Comment by cgadal** [comment]
This is because `numpy` has a safeguard in case `bin_min == bin_max`:

https://github.com/numpy/numpy/blob/9bfc8aaf70209df935e58c195b956f0a817e3755/numpy/lib/_histograms_impl.py#L320-L323

Adding something like:

```python
if bin_min == bin_max:
    bin_min = bin_min - 0.5
    bin_max = bin_max + 0.5
```

would probably solve the issue?

**Comment by guilhermeleobas** [comment]
Hi @cgadal, would you like to give it a try? The code for `np.histogram` is mostly on `arraymath.py`

**Comment by cgadal** [comment]
@guilhermeleobas Yes I can do it if there is no rush (incoming release or anything). 

I also have a local similar function that I based on the code that can be found on `arraymath.py` that extends the `weights` keyword and the `density` keyword, but currently only for regularly spaced bins.  We could also include this, and extend it for the irregularly spaced bins.
